### DISCUSSION TITLE
How to use middleware to add query parameter?

### DISCUSSION THREAD

**Discussion by forcia-tishii** [openapi-ts/openapi-typescript#1819] [Q&A]
I want to use a middleware to auth using an api key in query parameter. But, I can't use middleware to add a query parameter.

```ts
import createClient, { type Middleware } from 'openapi-fetch'
import type { paths } from './my-openapi-3-schema'

const middleware: Middleware = {
  async onRequest({ request }) {
    const url = new URL(request.url)
    const params = new URLSearchParams({ apiKey: 'foo' })
    url.search = params.toString()
    return new Request(url, {
      method: request.method,
      headers: request.headers,
      body: request.body
    })
  }
}

const client = createClient<paths>({ baseUrl: 'https://example.com/' })

client.use(middleware)

const { data, error } = await client.POST('/foo', {
  body: {
    foo: 'bar'
  }
})
```

Then, I got the following error

```console
TypeError: RequestInit: duplex option is required when sending a body.
    at new Request (node:internal/deps/undici/undici:5096:19)
    ...
```

**Comment by drwpow** [comment]
So this is a bug we’ve encountered on different implementations of `Request` in Node and browsers. Annoyingly, what you wrote _should_ work (at least as defined by the spec). But in practice, many implementations only support `new Request(options)` and will throw that error. To fix it, move `url` inside the object and only pass that one object to `Request`.

This isn’t a problem with this library; it’s a deviation in different implementations of how `Request` works across platforms.
